File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;
class Solution {
    Set<Integer> combinations = new HashSet<>();//비트마스킹으로 user_id인덱스스위치
    String[] user_id;
    String[] banned_id;
    public int solution(String[] user_id, String[] banned_id) {
        this.user_id = user_id;
        this.banned_id = banned_id;
        backtrack(0, 0, 0);
        return combinations.size();
    }
    //banned_id를 idx 증가하면서 탐색하며 user_id가 처리되었는지를 확인
    //처리안 된 유저에 한해 처리가능하다면(길이동일,*처리) 추가/추가안하기 백트래킹
    public void backtrack(int currentBit, int banned_id_idx, int bannedBit) {
        if (banned_id_idx == banned_id.length) {
            if (bannedBit != ((1 << banned_id.length) - 1)) return;//다안씀
            combinations.add(currentBit);
            return;
        }
        //banned_id_idx에 해당하는 banned_id에 한해서 user_id순회
        for (int i = 0; i < user_id.length; i++) {
            if ((currentBit & (1 << i)) != 0) continue;//이미처리된유저(못고름)
            if (matched(user_id[i], banned_id[banned_id_idx])) {
                backtrack(currentBit | (1 << i), banned_id_idx + 1, bannedBit | (1 << banned_id_idx));
            }
            //돌릴거없음 바뀐게없으니 다음 for턴에서 다른유저시도해볼거임
            //근데 아예 안쓰고 넘어가면 말안되는디.. 다써야만가능함..<췤~
        }
    }
    public boolean matched(String user, String ban) {
        if (user.length() != ban.length()) return false;
        int len = user.length();
        for (int i = 0; i < len; i++) {
            if (user.charAt(i) == ban.charAt(i)) continue;
            if (ban.charAt(i) == '*') continue;
            return false;
        }
        return true;
    }
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
import re 
def solution(user_id, banned_id):
    banToUser = []
    for ban in banned_id:
        reg = ban.replace('*', '.')
        matches = []
        for i, user in enumerate(user_id):
            if re.fullmatch(reg, user):
                matches.append(i)
        banToUser.append(matches)
        
    final_combinations = set()
    def helper(mask, idx):
        if idx == len(banToUser):
            final_combinations.add(mask)
            return
            
        userlists = banToUser[idx]
        for userIdx in userlists:
            if mask & (1 << userIdx): continue
            helper(mask | (1 << userIdx), idx + 1)
        
    helper(0,0)
    return len(final_combinations)

"""
우선 불량사용자에 해당할 수 있는 자료구조를 매핑한다(리스트)
그다음에 모든 경우의 수를 봐야하므로 조합을 해야한다
조합을 하려면 백트래킹을 해야한다
백트래킹하면서 유저아이디 사용여부(조합에 사용되었는지)를 확인한다. 이떄 비트마스킹을 사용한다
백트래킹의 종료조건은 매핑한 리스트의 엔딩에 도달하면이다. 이떄 set에 마스킹정보를 확인해서 경우의수를 확인한다.
"""